### PR TITLE
[lldb] Inline expression evaluator error visualization

### PR DESCRIPTION
**Pull request by adrian-prantl**
This patch is a reworking of Pete Lawrence's (@PortalPete) proposal
for better expression evaluator error messages:
https://github.com/llvm/llvm-project/pull/80938

Before:

```
$ lldb -o "expr a+b"
(lldb) expr a+b
error: <user expression 0>:1:1: use of undeclared identifier 'a'
a+b
^
error: <user expression 0>:1:3: use of undeclared identifier 'b'
a+b
  ^
```

After:

```
(lldb) expr a+b
            ^ ^
            │ ╰─ error: use of undeclared identifier 'b'
            ╰─ error: use of undeclared identifier 'a'
```

This eliminates the confusing `<user expression 0>:1:3` source
location and avoids echoing the expression to the console again, which
results in a cleaner presentation that makes it easier to grasp what's
going on. You can't see it here, bug the word "error" is now also in color, if so desired.

Depends on https://github.com/llvm/llvm-project/pull/106442.

### CODE REVIEW

File: lldb/source/Interpreter/CommandInterpreter.cpp [hunk 1]
```diff
@@ -2076,7 +2077,11 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
     }
 
     ElapsedTime elapsed(execute_time);
-    cmd_obj->Execute(remainder.c_str(), result);
+    size_t nchar = real_original_command_string.find(remainder);
+    std::optional<uint16_t> pos_in_cmd;
+    if (nchar != std::string::npos)
+      pos_in_cmd = nchar + GetDebugger().GetPrompt().size();
+    cmd_obj->Execute(remainder.c_str(), pos_in_cmd, result);
```
Comment (medismailben) [inline lldb/source/Interpreter/CommandInterpreter.cpp:2084]: Instead of updating every `DoExecute` implementations, it would be reasonable to have make this an optional member of the `CommandObject` base class so it can be accessed by the derived `CommandObject` classes that gonna use it.


The optional could be set in the non-virtual `CommandObject::Execute` command, since that's the one that calls the derived class `DoExecute` or may be by adding a setter to the `CommandObject` base class. 

This would greatly reduce the size of this patch.

**Review by medismailben — CHANGES_REQUESTED** [review]
Let's make `std::optional<uint16_t> offset_in_command` a default member of the `CommandObject` base class to avoid changing all the `DoExecute` implementations, even when not needed.

**Comment by adrian-prantl** [comment]
> This seems like it could be problematic for IDEs, if they don't print the error in the same window as the expression being evaluated. The arrows could end up pointing to nowhere, or to the wrong place in the expression (if we don't get the right offset).

Eventually I want IDEs to get access to the same kind of machine-readable diagnostics, so they can format errors however they like (or, if they aren't interested, dump the pre-rendered textual error that is still available).

> I think this could be made simpler and more robust by just printing a reproduction of the expression as the first line of the error message. It saving that one line output worth it?

Yes. It's not about saving one line of output, but for removing the cognitive load of having to re-read the expression, shifted by a couple of characters and correlating that with what you just typed, so you know where to make an edit.
 
> Also, how will this behave for multiline expressions? Or with errors that refer to multiple source locations (e.g inside macro expansions)?

I haven't wired that up correctly yet (thanks for reminding!), but the idea is that a multi-line expression would get the prerendered diagnostic as previously.

File: lldb/source/Interpreter/CommandInterpreter.cpp (hunk 1 above)
Comment (jimingham) [reply lldb/source/Interpreter/CommandInterpreter.cpp:2084]: I think expr is probably the only command that is likely to ever echo its command string into the output.  That's not something most commands would do.  So this doesn't seem like a feature that should have as much prominence as changing the API gives it.  I don't want to have to worry about what this means when I'm writing a new command, given the answer is almost surely to be "not relevant".

Comment (jimingham) [reply lldb/source/Interpreter/CommandInterpreter.cpp:2084]: Also, how does this work if I run

expr -i 0 -u 0 -- not a valid expression

The command string here is `-i 0 -u 0 -- not a valid expression` but you probably don't ever echo that full line onto the diagnostic, so either you are fudging and only matching the args part of the command, or the docs aren't right here.

Comment (jimingham) [reply lldb/source/Interpreter/CommandInterpreter.cpp:2084]: This seems a very expr specific function and not a generally useful for commands function.

Comment (medismailben) [reply lldb/source/Interpreter/CommandInterpreter.cpp:2084]: > I think expr is probably the only command that is likely to ever echo its command string into the output. That's not something most commands would do. So this doesn't seem like a feature that should have as much prominence as changing the API gives it. I don't want to have to worry about what this means when I'm writing a new command, given the answer is almost surely to be "not relevant".

I don't have a strong option about this (as long as we don't update all the `DoExecute` implementations), but I was under the impression that in the future, some commands could make use of it to point directly to what caused an error in their invocation. But again, we can have it be part of the expression for now.

**Comment by adrian-prantl** [comment]
Rebased on top of https://github.com/llvm/llvm-project/pull/106442!

File: lldb/source/Interpreter/CommandInterpreter.cpp (hunk 1 above)
Comment (adrian-prantl) [reply lldb/source/Interpreter/CommandInterpreter.cpp]: @jimingham @medismailben I hopefully have resolved both of your concerns by storing the original command string as a member in CommandObject. This way the `expr -i 0 -u 0 -- not a valid expression` works.
```
(lldb) expr -i 0 -u 0 -- not a valid expression
                         ^
                         ╰─ error: use of undeclared identifier 'not'
```

**Comment by jimingham** [comment]
Ah, I thought you were echoing something and pointing into that, I missed that you were pointing into the original command line.  Does this also work if I have an alias that scrambles the command arguments compared to how they are in the command that eventually got invoked.

If we're going to do errors this way, we should change the argument/option value setting to recognize which slot a bad option argument/option value came from (or unrecognized option flag for that matter) and point at those as well...

Jim

**Comment by adrian-prantl** [comment]
This is now ready for review. Out of scope for this PR, but the obvious next step: `dwim-print` / `p` support.

**Comment by adrian-prantl** [comment]
@medismailben @labath Would you mind taking another look at this?

File: lldb/source/Commands/CommandObjectExpression.cpp [hunk 2]
```diff
@@ -486,19 +603,37 @@ bool CommandObjectExpression::EvaluateExpression(llvm::StringRef expr,
...
+          bool multiline = expr.contains('\n');
+          RenderDiagnosticDetails(error_stream, expr_pos, multiline, details);
         } else {
-          error_stream.PutCString("error: unknown error\n");
+          const char *error_cstr = result_valobj_sp->GetError().AsCString();
+          if (error_cstr && error_cstr[0]) {
+            const size_t error_cstr_len = strlen(error_cstr);
+            const bool ends_with_newline =
+                error_cstr[error_cstr_len - 1] == '\n';
+            if (strstr(error_cstr, "error:") != error_cstr)
```
Comment (medismailben) [inline lldb/source/Commands/CommandObjectExpression.cpp:628]: Instead of using c-string you could wrap it in an `llvm::StringRef` and simplify this part here.

File: lldb/source/Interpreter/CommandInterpreter.cpp [hunk 3]
```diff
@@ -1887,7 +1887,8 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
                                        CommandReturnObject &result,
                                        bool force_repeat_command) {
   std::string command_string(command_line);
-  std::string original_command_string(command_line);
+  std::string original_command_string(command_string);
+  std::string real_original_command_string(command_string);
```
Comment (medismailben) [inline lldb/source/Interpreter/CommandInterpreter.cpp:1891]: nit: does this makes the variable above not the real original command string :p ? may be we can find a better name or write a comment explaining the nuance between the 2

File: lldb/source/Utility/Status.cpp [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 #include "lldb/Utility/Status.h"
 
+#include "lldb/Expression/DiagnosticManager.h"
```
Comment (medismailben) [inline lldb/source/Utility/Status.cpp:11]: Is this addition necessary ? It doesn't look like anything else has been added to this file

**Review by medismailben — APPROVED** [review]
This is pretty cool! LGTM with comment!

**Comment by labath** [comment]
> > This seems like it could be problematic for IDEs, if they don't print the error in the same window as the expression being evaluated. The arrows could end up pointing to nowhere, or to the wrong place in the expression (if we don't get the right offset).
> 
> Eventually I want IDEs to get access to the same kind of machine-readable diagnostics, so they can format errors however they like (or, if they aren't interested, dump the pre-rendered textual error that is still available).

I'm still worried about this. Yes, the IDEs can dump the pre-rendered error (and right now, that's all they can do), but this rendering assumes that the error message will be printed in a particular way (right under the original command, which will include the prompt and everything). I think that's fine for the commands entered through the LLDB command line, but I don't think it's reasonable to expect that every caller of `SBCommandInterpreter::HandleCommand` will do the same thing. I've looked at what lldb-dap that, and it looks like it will mostly be okay, because it explicitly echoes the command to be executed (although it [hardcodes the prompt string](https://github.com/llvm/llvm-project/blob/60ed2361c0917b4f8d54cb85935cfbf8904aa51d/lldb/tools/lldb-dap/LLDBUtils.cpp#L61) instead of getting it from lldb), but if you're looking for a example, you don't need to look further than your test case. Even though you've formatted the test input nicely, this is how its trace looks like when you run it:

```
Ran command:
"expr -i 0 -o 0 -- a"

Got output:
                         ^
                         ╰─ error: use of undeclared identifier 'a'
```

(sure we can fix this so that the output makes sense, but I wonder how many other such callers are out there)

**Comment by adrian-prantl** [comment]
> > > This seems like it could be problematic for IDEs, if they don't print the error in the same window as the expression being evaluated. The arrows could end up pointing to nowhere, or to the wrong place in the expression (if we don't get the right offset).
> > 
> > 
> > Eventually I want IDEs to get access to the same kind of machine-readable diagnostics, so they can format errors however they like (or, if they aren't interested, dump the pre-rendered textual error that is still available).
> 
> I'm still worried about this. Yes, the IDEs can dump the pre-rendered error (and right now, that's all they can do), but this rendering assumes that the error message will be printed in a particular way (right under the original command, which will include the prompt and everything). I think that's fine for the commands entered through the LLDB command line, but I don't think it's reasonable to expect that every caller of `SBCommandInterpreter::HandleCommand` will do the same thing. I've looked at what lldb-dap does, and it looks like it will mostly be okay, because it explicitly echoes the command to be executed (although it [hardcodes the prompt string](https://github.com/llvm/llvm-project/blob/60ed2361c0917b4f8d54cb85935cfbf8904aa51d/lldb/tools/lldb-dap/LLDBUtils.cpp#L61) instead of getting it from lldb), but if you're looking for a example, you don't need to look further than your test case. Even though you've formatted the test input nicely, this is how its trace looks like when you run it:
> 
> ```
> Ran command:
> "expr -i 0 -o 0 -- a"
> 
> Got output:
>                          ^
>                          ╰─ error: use of undeclared identifier 'a'
> ```
> 
> (sure we can fix this so that the output makes sense, but I wonder how many other such callers are out there)

Just to explain my motivation., in the end what I'm most interested in is exposing the machine-readable diagnostics through the SBAPI, so that is where this is going next.
I'm actually surprised that lldb-dap sees the new diagnostics because I thought it would run the expression through a lower-level API like SBFrame::EvaluateExpression(). But I guess it just provides a "terminal" window that passes everything through HandleCommand. To support this better, I'm thinking about adding a setting to opt into the inline diagnostics that only `lldb` sets, so we don't get unexpected results like this.

**Comment by adrian-prantl** [comment]
Done. @labath, you can now opt into the inline diagnostics with `show-inline-diagnostics` which `tools/driver` sets.

File: lldb/source/Core/Debugger.cpp [hunk 5]
```diff
@@ -592,7 +592,18 @@ lldb::DWIMPrintVerbosity Debugger::GetDWIMPrintVerbosity() const {
   const uint32_t idx = ePropertyDWIMPrintVerbosity;
   return GetPropertyAtIndexAs<lldb::DWIMPrintVerbosity>(
       idx, static_cast<lldb::DWIMPrintVerbosity>(
-               g_debugger_properties[idx].default_uint_value));
+               g_debugger_properties[idx].default_uint_value != 0));
```
Comment (labath) [inline lldb/source/Core/Debugger.cpp:595]: Did you intend to change DWIM code?

**Review by labath — APPROVED** [review]
The setting might be useful as well, though I think it would be even better if this was settable on a per-invocation basis (an argument to HandleCommand or something) -- the reason being that someone might still want to have this feature for commands which are input through the console (or from contexts where you can guarantee their placement), but not for commands which are executed e.g. from scripts. Here's one example:
```
(lldb) script lldb.debugger.HandleCommand('expression -- ""+2.5')
error: <user expression 0>:1:3: invalid operands to binary expression ('const char[1]' and 'double')
    1 | ""+2.5
      | ~~^~~~
```
You definitely won't point to the right place there. It's convoluted, I know, but it's not that unsimilar from I do when debugging lldb, where I sometimes run things like `self.dbg.HandleCommand` from the pdb prompt.

However this is workaroundable by changing the setting, so I don't want to hold this up further.

**Comment by labath** [comment]
> I'm actually surprised that lldb-dap sees the new diagnostics because I thought it would run the expression through a lower-level API like SBFrame::EvaluateExpression(). But I guess it just provides a "terminal" window that passes everything through HandleCommand. To support this better, I'm thinking about adding a setting to opt into the inline diagnostics that only `lldb` sets, so we don't get unexpected results like this.

It has both. lldb-dap does this weird heuristic console multiplexing, where it guesses whether something should be treated like an expression (I guess that's because other VSCode debuggers do that), which is executed through EvaluateExpression, or an lldb command, which is executed through HandleCommand. So it's not very likely that a user will be running a "command" to evaluate an expression, but I think it still illustrates my point that HandleCommand can be (and probably is) invoked in contexts where one cannot guarantee the alignment of the output.

**Comment by jimingham** [comment]
This is a slightly orthogonal point, but I don't think we should confine this error visualization to just expressions.  If the user passes a non-boolean string to an option in a command, it would be helpful to point at the option for them, much as pointing to the part of the expression that doesn't parse is useful.  And if we're pointing at errors in expressions, it looks kind of odd that we don't do the same thing for other commands.

I'm not suggesting that that be part of this work, but when thinking about when this kind of thing would be appropriate I think would be unfortunate if we didn't allow the same feature at least for parsed commands.

Jim

File: lldb/source/Interpreter/CommandInterpreter.cpp (hunk 3 above)
Comment (adrian-prantl) [reply lldb/source/Interpreter/CommandInterpreter.cpp:1891]: This is just me being funny. I thought that if there is a std::string implementation that does CoW, then this would be more efficient, since the string data could be shared.
